README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
resurrectionr <img src="man/figures/logo.png" align="right" />
==============================================================

[![Travis-CI Build
Status](https://travis-ci.org/mdjeric/resurrectionr.svg?branch=master)](https://travis-ci.org/mdjeric/resurrectionr)

The purpose of resurrectionr is `recode_religion()` which recodes a set
of religious identification variables from the US General Social Survey,
in a way that is analytically useful and easy to implement in data
processing.

Bigger sociological and political problems in recoding of religious
identification are elaborated and resolved with a recoding schema in the
paper by Darren E. Sherkat and Derek Lehman,[1] based on
classification[2] from 2001. This package provides solutions for main
technical issues with recoding of religious identification from GSS in
R, using said schema.

1.  Variables contain large number of codes: 16 (`relig`), 30 (`denom`),
    and 201 (`other`).  
2.  Same pattern of variables describes religious identification for a
    large number of items (e.g. spouses’, father’s, mother’s, etc.), and
    all of them have, naturally, different variable names.  
3.  Punch codes are not consequtive indexes, but are often skipping
    certain numbers, and, for certain items they tend to have only
    punches, not labels.

This package provides universal function `recode_religion()` that can
recode all these varaibles and adds additional benefits in terms of
cheking values, providing a recoding key, etc.

Installation
------------

``` r
# Install development version from GitHub
# install.packages("devtools") # If you don't have it
devtools::install_github("mdjeric/resurrectionr")
```

Example
-------

This is a basic use of function:

``` r
library(resurrectionr)
gss <- gss14_f


gss$religion <- recode_religion(gss$relig, gss$denom, gss$other)
#> Distribution of religious identification, in your data of 2538 is:
#>                           Freq Relative  Cumul
#> Catholic or Orthodox       615    24.23  24.23
#> None                       522    20.57  44.80
#> Baptist                    324    12.77  57.57
#> Christian, no group given  307    12.10  69.66
#> Moderate Protestant        206     8.12  77.78
#> Sectarian Protestant       178     7.01  84.79
#> Lutheran                    92     3.62  88.42
#> Other religion              88     3.47  91.88
#> Liberal Protestant          77     3.03  94.92
#> Jewish                      40     1.58  96.49
#> Episcopalian                39     1.54  98.03
#> Mormon                      32     1.26  99.29
#> No answer                   15     0.59  99.88
#> Don't know                   3     0.12 100.00


summary(gss$religion)
#>                   Baptist      Catholic or Orthodox 
#>                       324                       615 
#> Christian, no group given                Don't know 
#>                       307                         3 
#>              Episcopalian                    Jewish 
#>                        39                        40 
#>        Liberal Protestant                  Lutheran 
#>                        77                        92 
#>       Moderate Protestant                    Mormon 
#>                       206                        32 
#>                 No answer                      None 
#>                        15                       522 
#>            Other religion      Sectarian Protestant 
#>                        88                       178
```

Description
-----------

`recode_religion()` is a wrapper for basic function that does all the
recoding. It confirms the variables are with proper codes, same length,
transforms the type, and provides additional information, warning
messages, errors, etc.

`fct_rec_relig()`, the simple function, uses data frames in the list
[`rdo_cdbk`](reference/rdo_cdbk.html) from which three vectors are
formed where index corresponds to punch number, and label is value (of
`relig`, `denom`, and `other`).

For each *new* religious identification, separate vectors (for `relig`,
`denom`, and/or `other` identification) are created that contain
corresponding punch codes and labels, per paper and SPSS syntax. In
addition, logical vectors are created for Sectarian Protestants with
valid denomination codes for sorting between ‘Sectarian Protestants’,
‘Christian - no group given’, and ‘other religions’.

Next, twelve logical vectors, for respondent’s belonging to each group,
are created by checking against appropriate vectors containing labels.
Names are assigned, function prints frequency table, and returns vector
that contains new variable.

Additional behavior of the function
-----------------------------------

Function checks that the (1) vectors are of same length, and (2) that
their values are valid answers, throwing errors like this:

``` r
religion <- recode_religion("PROTESTANT", c(12, 13), "IAP")
#> Error: Vectors must be of the same lenght, currently they are:
#> Length of `relig`: 1.
#> Length of `denom`: 2.
#> Length of `other`: 1.
```

It is also possible to combine punches with labels, and get same
results, or print the coding key that was used on specific data.

``` r
set.seed(999) # for reproducability
short <- sample(1:2358, 50, replace = FALSE) # chose random 50 cases
gshrt <- gss[short, ]


gshrt$religion <- recode_religion(gshrt$relig, gshrt$denom, gshrt$other,
                                  add_missing_levels = TRUE, print_key = TRUE)
#> Distribution of religious identification, in your data of 50 is:
#>                           Freq Relative Cumul
#> Christian, no group given   13       26    26
#> Catholic or Orthodox        12       24    50
#> None                         7       14    64
#> Moderate Protestant          6       12    76
#> Baptist                      3        6    82
#> Liberal Protestant           2        4    86
#> Lutheran                     2        4    90
#> Other religion               2        4    94
#> Sectarian Protestant         2        4    98
#> Jewish                       1        2   100
#> Don't know                   0        0   100
#> No answer                    0        0   100
#> Episcopalian                 0        0   100
#> Mormon                       0        0   100
#> *** Key for recoding variables ***
#>  In this case, number of combinations was: 19 .
#>                                                                  
#> religion|           Baptist          Baptist Catholic or Orthodox
#>    ^    +              ^^^              ^^^                  ^^^ 
#>  relig  |        PROTESTANT       PROTESTANT             CATHOLIC
#>  denom  |  BAPTIST-DK WHICH SOUTHERN BAPTIST                  IAP
#>  other  |               IAP              IAP                  IAP
#> --------+-       ----------       ----------           ----------
#>                                                                          
#> religion|  Christian, no group given Christian, no group given     Jewish
#>    ^    +                       ^^^                       ^^^        ^^^ 
#>  relig  |                  CHRISTIAN                PROTESTANT     JEWISH
#>  denom  |            NO DENOMINATION           NO DENOMINATION        IAP
#>  other  |                        IAP                       IAP        IAP
#> --------+-                ----------                ---------- ----------
#>                                                                    
#> religion|  Liberal Protestant   Liberal Protestant         Lutheran
#>    ^    +                ^^^                  ^^^              ^^^ 
#>  relig  |          PROTESTANT           PROTESTANT       PROTESTANT
#>  denom  |  PRESBYTERIAN-DK WH PRESBYTERIAN, MERGED EVANGELICAL LUTH
#>  other  |                 IAP                  IAP              IAP
#> --------+-         ----------           ----------       ----------
#>                                                                  
#> religion|        Lutheran Moderate Protestant Moderate Protestant
#>    ^    +            ^^^                 ^^^                 ^^^ 
#>  relig  |      PROTESTANT          PROTESTANT          PROTESTANT
#>  denom  |  OTHER LUTHERAN   AM BAPT CH IN USA   NAT BAPT CONV USA
#>  other  |             IAP                 IAP                 IAP
#> --------+-     ----------          ----------          ----------
#>                                                                       
#> religion|  Moderate Protestant Moderate Protestant Moderate Protestant
#>    ^    +                 ^^^                 ^^^                 ^^^ 
#>  relig  |           PROTESTANT          PROTESTANT          PROTESTANT
#>  denom  |                OTHER               OTHER    UNITED METHODIST
#>  other  |     Christian Reform Disciples of Christ                 IAP
#> --------+-          ----------          ----------          ----------
#>                                                             
#> religion|        None Other religion          Other religion
#>    ^    +        ^^^            ^^^                     ^^^ 
#>  relig  |        NONE       HINDUISM              PROTESTANT
#>  denom  |         IAP            IAP                   OTHER
#>  other  |         IAP            IAP Unitarian, Universalist
#> --------+- ----------     ----------              ----------
#>                                        
#> religion|          Sectarian Protestant
#>    ^    +                          ^^^ 
#>  relig  |                    PROTESTANT
#>  denom  |                         OTHER
#>  other  |  Christian; Central Christian
#> --------+-                   ----------
#> ===================================
#> This is probably not the best way to present or inspect them.
#>  There will be a better way soon, or use unique() on your data.
```

It also works well when some (or all) variables contain punches, either
as characters or numbers.

``` r
gshrt$d_num <- gss14_n$denom[short]
gshrt$o_num_char <- as.character(gss14_n$other[short])


gshrt$religion2 <- recode_religion(gshrt$relig, gshrt$d_num, gshrt$o_num_char,
                                  add_missing_levels = TRUE, frequencies = FALSE)
#> * `denom` recoded from punches to labels; and 'NA' introduced.
#> * `other` recoded from punches to labels; and 'NA' introduced.

# We can see that there is no difference
FALSE %in% (gshrt$religion2 == gshrt$religion)
#> [1] FALSE
```

Bugs
----

There are minor differences, however, when all missing values are
imported from SPSS. For example, if all missing values are treated
equaliy, certain groups might appear `NA`, as seen in this example:

``` r
gss$religion2 <- recode_religion(gss14_n$relig, gss14_n$denom, gss14_n$other, frequencies = FALSE)
#> Some of the variables contain NA: `Don't know` and `NA`will be merged. Please see documentation for more details.
#> * `relig` recoded from punches to labels; and 'NA' introduced.
#> * `denom` recoded from punches to labels; and 'NA' introduced.
#> * `other` recoded from punches to labels; and 'NA' introduced.
unique(gss[,c("religion2", "religion")])
#>                     religion2                  religion
#> 1        Catholic or Orthodox      Catholic or Orthodox
#> 3                    Lutheran                  Lutheran
#> 6                Episcopalian              Episcopalian
#> 13  Christian, no group given Christian, no group given
#> 18                       None                      None
#> 19        Moderate Protestant       Moderate Protestant
#> 21         Liberal Protestant        Liberal Protestant
#> 34             Other religion            Other religion
#> 43       Sectarian Protestant      Sectarian Protestant
#> 46                    Baptist                   Baptist
#> 54                     Jewish                    Jewish
#> 119                      <NA>                 No answer
#> 220                      <NA>                Don't know
#> 260                    Mormon                    Mormon
#> 309                      <NA> Christian, no group given

# total number of cases in 2014 GSS:
gss[is.na(gss$religion2) & gss$religion == "Christian, no group given",]
#>               bible      relig denom other                  religion
#> 309   INSPIRED WORD PROTESTANT OTHER    NA Christian, no group given
#> 419     WORD OF GOD PROTESTANT    DK   IAP Christian, no group given
#> 677   INSPIRED WORD PROTESTANT OTHER    DK Christian, no group given
#> 1138 BOOK OF FABLES PROTESTANT    DK   IAP Christian, no group given
#> 2487    WORD OF GOD PROTESTANT    DK   IAP Christian, no group given
#>      religion2
#> 309       <NA>
#> 419       <NA>
#> 677       <NA>
#> 1138      <NA>
#> 2487      <NA>
# These are with different types of missing values on denom and other
# should be corrected soon.
```

Further work
------------

Three more things are planned to be included in this package:

1.  Recode to often more useful group of 7 religious identifications.
2.  Function that just prints the results.
3.  Function that returns key with all recoding tetrades, and all
    recoding tetrades in particular set, as data frame (it might be
    useful for someone).

Code of conduct
---------------

Please note that this project is released with a [Contributor Code of
Conduct](CONDUCT.md). By participating in this project you agree to
abide by its terms.

[1] <a href="https://iranianredneck.wordpress.com/2016/11/29/why-reltrad-sucks-contesting-the-measure-of-american-religion/" target="_blank">“After The Resurrection: The Field of the Sociology of Religion in the United States”</a>

[2] Sherkat, Darren E. 2001. “Tracking the restructuring of American
religion: Religious affiliation and patterns of religious mobility,
1973–1998.” *Social Forces* 79(4), 1459-1493.
[doi:10.1353/sof.2001.0052](https://doi.org/10.1353/sof.2001.0052)


NEWS.md
# resurrectionr 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.
* Created initial package from the old function.


R/data.R
#' Codebook for relig, denom, and other variables.
#'
#' A list containing data frames that have individual punches and
#' labels for \code{relig}, \code{denom}, and \code{other} variables, which are also
#' applicable for whole family of variables such as religious
#' identification at the age of 16, spouses, parents, etc. in General
#' Social Survey from 1972 until 2014. Some of these non-primariy
#' variables are often coded only with punches, and not with labels.
#' It's used in \code{\link{rdo_punches}}.
#'
#' @format A list containing three data frames:
#' \describe{
#'   \item{relig}{data frame with 16 observations and 2 variables:
#'   \code{punch}, numeric; \code{label} factor.}
#'   \item{denom}{data frame with 30 observations and 2 variables:
#'   \code{punch}, numeric; \code{label} factor.}
#'   \item{other}{data frame with 201 observation and 2 variables:
#'   \code{punch}, numeric; \code{label} factor.}
#' }
#' @source \href{http://www.gss.norc.org/Get-Documentation}{General Social Survey}
"rdo_cdbk"

#' GSS 2014 select variables as factors.
#'
#' Four varaibles from US Genearl Social Survey from 2014, three describe
#' respondent's religious identification(\code{relig}, \code{denom}, \code{other}),
#' and fourth \code{bible} is for biblical litteracy, or answers on the question
#' if the bible is "the word of god", "inspired word", "book of fables", or "other".
#' Data is used in vignettes, to demonstrate the use of
#' \code{\link{recode_religion}}, and to provide some comparison between
#' the use of packages \code{foreign} and \code{haven} in importing SPSS data
#' and their different recoding of categorical variables from GSS. \code{gss14_f}
#' was imported through \code{foreign}, with use of following options:
#' \code{to.data.frame = TRUE, trim.factor.names = TRUE, trim_values = TRUE,
#' use.missings = FALSE}, accordingly varaibles are factors without missing
#' values.
#'
#' @seealso \code{\link[foreign]{read.spss}} \code{\link[haven]{read_sav}}
#'
#' @format
#' \describe{A data frame with 2538 rows and 4 variables:
#'   \item{bible}{Factor, feelings about the bible.}
#'   \item{relig}{Factor, respondents religious preference.}
#'   \item{denom}{Factor, specific denomination.}
#'   \item{other}{Factor, other Protestant denominations.}
#' }
#'
#' @source \href{http://www.gss.norc.org/get-the-data/spss}{GSS 2014}
"gss14_f"

#' GSS 2014 select variables as numerical.
#'
#' Four varaibles from US Genearl Social Survey from 2014, three describe
#' respondent's religious identification(\code{relig}, \code{denom}, \code{other}),
#' and fourth \code{bible} is for biblical litteracy, or answers on the question
#' if the bible is "the word of god", "inspired word", "book of fables", or "other".
#' Data is used in vignettes, to demonstrate the use of
#' \code{\link{recode_religion}}, and to provide some comparison between
#' the use of packages \code{foreign} and \code{haven} in importing SPSS data
#' and their different recoding of categorical variables from GSS. \code{gss14_n}
#' was imported through \code{haven}, and thus variables are specific numeric
#' \code{labelled} variables, with attributes "label" for description,
#' "format.spss" for SPSS format, and "labels" which are names for levels.
#'
#' @seealso \code{\link[haven]{read_sav}} \code{\link[foreign]{read.spss}}
#'
#' @format
#' \describe{A data frame with 2538 rows and 4 variables:
#'   \item{bible}{Labelled num, feelings about the bible.}
#'   \item{relig}{Labelled num, respondents religious preference.}
#'   \item{denom}{Labelled num, specific denomination.}
#'   \item{other}{Labelled num, other Protestant denominations.}
#' }
#'
#' @source \href{http://www.gss.norc.org/get-the-data/spss}{GSS 2014}
"gss14_n"


R/recode_religion.R
#' Recode Religion from GSS.
#'
#' Function recodes religious identification from Genearl Social Survey,
#' based on three variables: \code{relig}, \code{denom}, and \code{other}.
#' It can successfully recode either respondent's, or any other religious
#' identification which is determened by coresponding three variables.
#'
#' \code{recode_religion} uses schema developed by Darren E. Sherkat and
#' Derek Lehman in \href{https://iranianredneck.wordpress.com/2016/11/29/why-reltrad-sucks-contesting-the-measure-of-american-religion/}{"After
#' The Resurrection: The Field of the Sociology of Religion in the United
#' States"}, and is effectievly translation of that SPSS syntax (the bare
#' bone function for recoding is \link{fct_rec_relig}), with
#' additional functionality.
#'
#' Namley, it can handle both punches and labels
#' at the same time (but in different variables), which is important since
#' punches are not consequtive as indexes. In addition, function checks that
#' variables are adequate (i.e. that all values are in codebook) and of same
#' length, and also handles missing values: (1) if supplied through values,
#' provides detail recoding; (2) if \code{NA}, lumps them together in final
#' variable but uses them correctly in the recoding. Through passed
#' arguments, one can:
#' \enumerate{
#'   \item Return longer version, of 12 religious identifications.
#'   \item Add identifications from schema that are not present in
#'   sample as empty levels.
#'   \item Suppress printing of the frequencies of newly recoded variable.
#'   \item Print unique key of the values that were recoded.
#'   \item Return values as numerical factor, in which case the codebook
#'   for new variable will be printed.
#' }
#' If \code{frequencies} is passed as \code{FALSE}, and numerical vector
#' is not requested as return value, all other information, such as
#' treatment of missing values, are provided as messages that can be
#' suppressed.
#'
#' More details can be found on \href{https://github.mdjeric}{github.mdjeric}.
#'
#' @param relig,denom,other Numerical, character, or factor, all of same
#' length and with coresponding punches or labels in codebook.
#' @param n_groups Number, 7 (default) or 12 of new religious identifications.
#' @param add_missing_levels Logical, to include as empty levels religious
#' identifications that may not be present in specific sample, but are
#' part of recoding schema.
#' @param frequencies Logical, to print frequency and percent table of
#' recoded religius identification (default is \code{TRUE}).
#' @param print_key Logical, to print the all unique tetrads of recoded
#' variables, i.e. a recoding key.
#' @param return_num Logical, to return numerical factor and print codebook.
#'
#' @return Vector with recoded religion from \code{relig}, \code{denom},
#' and \code{other}. Function does not return \code{NA}, but as
#' factor levels "Not answered" and "Don't know", or combined
#' "Not answered/Don't know" when missing values are not
#' declared as punches or labels in initial variables but passed on as
#' \code{NA} (function gives message and where \code{NAs} are lcoated).
#' Default is to have factor with 7 descriptive levels, but function can
#' also return numerical vector. Default behavior returns only present
#' values, but can be made to add additional empty levels if \code{TRUE}
#' is passed to \code{add_missing_levels}. Function also \strong{prints}
#' frequency table of newly recoded religious identification, which can be
#' suppressed with \code{frequencies}. If required, it can also return
#' numerical vector and print the coding for it (not recommended).
#'
#' @examples
#' library(resurrectionr)
#'
#' # When all variables are factor
#' gss14_f$religion <- recode_religion(gss14_f$relig, gss14_f$denom,
#'                                     gss14_f$other, frequencies = FALSE)
#' # Twelve groups
#' gss14_f$religion <- recode_religion(gss14_f$relig, gss14_f$denom,
#'                                     gss14_f$other, n_groups = 12)
#'
#' # When all variables are numeric
#' gss14_n$religion <- recode_religion(gss14_n$relig, gss14_n$denom,
#'                                     gss14_n$other,
#'                                     add_missing_levels = TRUE)
#'
#' # But also, combining them works
#' religion <- recode_religion(gss14_f$relig, gss14_n$denom,
#'                             as.character(gss14_f$other))
#' @export
recode_religion <- function(relig, denom, other, n_groups = 7,
                            add_missing_levels = FALSE, frequencies = TRUE,
                            print_key = FALSE, return_num = FALSE)  {

# Check that all arguments are of appropriate type ------------------------

  arg_err <- list(error = FALSE, name = c(), type = c())

  if (!(is.numeric(relig) | is.factor(relig) | is.character(relig)))
    arg_err <- add_error(arg_err, "relig",
                         "vector (numeric or character) or factor")

  if (!(is.numeric(denom) | is.factor(denom) | is.character(denom)))
    arg_err <- add_error(arg_err, "denom",
                         "vector (numeric or character) or factor")

  if (!(is.numeric(other) | is.factor(other) | is.character(other)))
    arg_err <- add_error(arg_err, "other",
                         "vector (numeric or character) or factor")

  if (!(as.character(n_groups) %in% c("7", "12")))
    arg_err <- add_error(arg_err, "n_groups", "7 or 12")

  if (!is.logical(add_missing_levels))
    arg_err <- add_error(arg_err, "add_missing_levels", "logical")

  if (!is.logical(frequencies))
    arg_err <- add_error(arg_err, "frequencies", "logical")

  if (!is.logical(print_key))
    arg_err <- add_error(arg_err, "print_key", "logical")

  if (!is.logical(return_num))
    arg_err <- add_error(arg_err, "return_num", "logical")

  # stop if one or more arguments are mismatched
  if (arg_err$error) stop("Arguments are not of appropriate type:\n",
                           sprintf(" * `%s` must be %s.\n",
                                   arg_err$name, arg_err$type),
                           call. = FALSE
                          )

# Check vector lengths ----------------------------------------------------

  if ((length(relig) != length(denom)) |
      (length(relig) != length(other)) |
      (length(denom) != length(other)) )
    stop("Vectors must be of the same lenght, currently they are:\n",
         "* Length of `relig`: ", length(relig), ".\n",
         "* Length of `denom`: ", length(denom), ".\n",
         "* Length of `other`: ", length(other), ".\n",
         call. = FALSE
         )

# Check for NA and transform them -----------------------------------------

  merge_na_dk <- FALSE
  if (any(is.na(relig)) & any(is.na(denom)) & any(is.na(other)))  {
    message("Some of the variables contain NA: `Don't know` and `NA`",
            "will be merged. Please see documentation for more details.")
    merge_na_dk <- TRUE
  }

  relig <- transform_rdo(relig, "relig")
  denom <- transform_rdo(denom, "denom")
  other <- transform_rdo(other, "other")


# Check that there are no unallowed values --------------------------------

  error_values <- list(error = FALSE, name = c(), type = c())
  if (is.logical(relig[[1]])) error_values <- add_error(error_values,
                                                        relig[[2]],
                                                        relig[[3]])
  if (is.logical(denom[[1]])) error_values <- add_error(error_values,
                                                        denom[[2]],
                                                        denom[[3]])
  if (is.logical(other[[1]])) error_values <- add_error(error_values,
                                                        other[[2]],
                                                        other[[3]])

  if (error_values$error) {
    stop("Variables with values that are not in codebook:\n",
         sprintf(" * `%s` has, for example: %s.\n",
                 error_values$name, error_values$type),
         call. = FALSE)
  }



# Create recoded religion -------------------------------------------------

  religion <- fct_rec_relig(relig, denom, other)

  if (merge_na_dk)  {
    religion[religion == "Don't know"] <- "Don't know/No answer"
    religion[religion == "No answer"] <- "Don't know/No answer"
  }

# Reduce groups if needed -------------------------------------------------
  all_levels <- c("Sectarian Protestant", "Baptist",
                  "Moderate Protestant", "Christian, no group given", "Lutheran",
                  "Liberal Protestant", "Episcopalian",
                  "Catholic and Orthodox",
                  "Other religions", "Jew",
                  "Mormon",
                  "No identification"
                  )

  if (as.character(n_groups) == "7")  {

    names(all_levels) <- c(1, 11, 2, 21, 22, 3, 31, 4, 5, 51, 6, 7)

    religion[religion %in% all_levels[c("1", "11")]] <- all_levels["1"]
    religion[religion %in% all_levels[c("2", "21", "22")]] <- all_levels["2"]
    religion[religion %in% all_levels[c("3", "31")]] <- all_levels["3"]
    religion[religion %in% all_levels[c("5", "51")]] <- all_levels["5"]

    all_levels <- unname(all_levels[as.character(c(1:7))])

  }

  religion <- as.factor(religion)

# Add missing levels ------------------------------------------------------

  if (add_missing_levels)  {
    missing_levels <- !(all_levels %in% levels(religion))
    if (TRUE %in% missing_levels)
      levels(religion) <- c(levels(religion), all_levels[missing_levels])
  }


# Print frequencies -------------------------------------------------------

  if (frequencies) print_frequencies(religion)


# Print recoding key ------------------------------------------------------

  if (print_key)  {
    DF <- data.frame('relig' = relig,
                     'denom' = denom,
                     'other' = other,
                     'religion' = religion
                     )
    print_key(DF)
  }


# Change to numeric -------------------------------------------------------

  if (return_num)  {
    indeksi <- 1:length(levels(religion))
    message("ATTENTION: Returned is numeric vector without attributes, ",
            "and with follwoing coressponding values: ",
            sprintf("[%i] %s * ", indeksi, levels(religion)))
    religion <- as.numeric(religion)
  }


# Return vector -----------------------------------------------------------

  return(religion)

}


R/fct_rec_relig.R
#' Basic religion recoding.
#'
#' @description With correct parameters, it effectievly recodes
#' the religion variables. \code{\link{recode_religion}} wraps it,
#' and conducts series of checks to confirm that variables for
#' recoding are good and that it will produce a correct religious
#' identification. In addition, \code{\link{recode_religion}} allows
#' for more flexibility in terms of output. However, this function
#' is the essence of the package, and contains details about logic
#' of the recoding schema, and can also be inspected at
#' \href{https://github.com/mdjeric/resurrectionr/R/recode_religion_only-factors.R}{github.com/mdjeric/resurrectionr}
#'
#' @param religion,denomination,other Factors.
#'
#' @return Character vector with recoded religion.
#'
fct_rec_relig <- function(religion, denomination, other)  {
  ######## Recoding religion and religion at 16 ##################
  # Sherkat and Lehman (2017)
  # To work properly, folder 'religion_punches' with .csvs
  # of label names has to be in wokring directory.
  #
  # Function: # rec_relig_12(religion, denomination, other)
  #	relig or relig16 variable; denom or denom 16; other or oth16
  # function prints frequencies and returns factor vector with
  # religion recoded.
  #
  # It works with GSS dataset imported through 'read.spss',
  # from foreign package, in following way:
  # to.data.frame = TRUE, trim.factor.names = TRUE,
  # trim_values = TRUE, use.missings = FALSE


  ## SETTING UP ==================================================

  # Import three varaibles into new dataset used for recoding ----
  DF <- data.frame(relig = religion,
                   denom = denomination,
                   other = other
                   )

  # Create vectors with position corespondign to the punch
  # of label in DF codebook for 3 variables
  c_r <- rdo_punches("relig")
  c_d <- rdo_punches("denom")
  c_o <- rdo_punches("other")


  ## RECODING =====================================================

  # Liberal Protestants --------------------------------------------
  lp_d_num <- c(40:49)
  lp_o_num <- c(29, 30, 40, 54, 70, 72 , 81, 82, 95, 98, 119,
                142, 160, 188)
  lp_denom <- c_d[lp_d_num]
  lp_other <- c_o[lp_o_num]
  DF$lp_true <- (DF$denom %in% lp_denom) | (DF$other %in% lp_other)
  DF$rv[DF$lp_true] <- "Liberal Protestant"

  # Episcopalians --------------------------------------------------
  ep_d_num <- c(50)
  ep_denom <- c_d[ep_d_num]
  DF$ep_true <- DF$denom %in% ep_denom
  DF$rv[DF$ep_true] <- "Episcopalian"

  # Moderate Protestants ------------------------------------------
  mp_d_num <- c(10:13, 20:23, 28)
  mp_o_num <- c(1, 8, 15, 19, 25, 32, 42:44, 46, 49:51, 71, 73, 94,
                99, 146, 148, 150, 186)
  mp_denom <- c_d[mp_d_num]
  mp_other <- c_o[mp_o_num]
  DF$mp_true <- (DF$denom %in% mp_denom) | (DF$other %in% mp_other)
  DF$rv[DF$mp_true] <- "Moderate Protestant"

  # Lutherans -----------------------------------------------------
  lt_d_num <- c(30:38)
  lt_o_num <- c(105)
  lt_denom <- c_d[lt_d_num]
  lt_other <- c_o[lt_o_num]
  DF$lt_true <- (DF$denom %in% lt_denom) | (DF$other %in% lt_other)
  DF$rv[DF$lt_true] <- "Lutheran"

  # Baptists ------------------------------------------------------
  bp_d_num <- c(14:18)
  bp_o_num <- c(93, 133, 197)
  bp_denom <- c_d[bp_d_num]
  bp_other <- c_o[bp_o_num]
  DF$bp_true <- (DF$denom %in% bp_denom) | (DF$other %in% bp_other)
  DF$rv[DF$bp_true] <- "Baptist"

  # Sectarian Protestants -----------------------------------------
  # these initial variables pull out sectarians codes
  # relig == 11 (christian) or relig == 5 (other),
  # but also have valid denom codes.
  DF$sp_pent <- (DF$relig == c_r[11]) & (DF$other == c_o[68])
  DF$sp_centchrist <- (DF$relig == c_r[5]) & (DF$other == c_o[31])
  DF$sp_fsg <- (DF$relig == c_r[5]) & (DF$other == c_o[53])
  DF$sp_jw <- (DF$relig == c_r[5]) & (DF$other == c_o[58])
  DF$sp_sda <- (DF$relig == c_r[5]) & (DF$other == c_o[77])
  DF$sp_ofund <- (DF$relig == c_r[5]) & (DF$other == c_o[97])

  sp_o_num <- c(2, 3, 5:7, 9, 10, 12:14, 16:18, 20:24, 26, 27, 31,
                33:39, 41, 45, 47, 48, 52, 53, 55:58, 63, 65:69,
                76:79, 83:92, 96, 97, 100:104, 106:113, 115:118,
                120:122, 124, 125, 127:132, 134, 135, 137:141, 144,
                145, 151:156, 158, 159, 166:182, 184, 185, 187,
                189:191, 193, 195, 196, 198, 201, 204)
  sp_other <- c_o[sp_o_num]

  DF$sp_true <- ((DF$other %in% sp_other) |
                  DF$sp_pent |
                  DF$sp_centchrist |
                  DF$sp_fsg |
                  DF$sp_jw |
                  DF$sp_sda |
                  DF$sp_ofund
                 )
  DF$rv[DF$sp_true] <- "Sectarian Protestant"

  # Christian, no group identified --------------------------------
  DF$cn_christ <- (DF$relig == c_r[11]) & !DF$sp_pent
  cn_r_num <- c(13)
  cn_d_num <- c(70, 98, 99)
  cn_o_num <- c(998, 999)
  cn_relig <- c_r[cn_r_num]
  cn_denom <- c_d[cn_d_num]
  cn_other <- c_o[cn_o_num]
  DF$cn_true <- ((DF$relig %in% cn_relig) |
                 (DF$denom %in% cn_denom) |
                 (DF$other %in% cn_other) |
                  DF$cn_christ
                 )
  DF$rv[DF$cn_true] <- "Christian, no group given"

  # Mormons -------------------------------------------------------
  mr_o_num <- c(59:62, 64, 157, 162)
  mr_other <- c_o[mr_o_num]
  DF$mr_true <- DF$other %in% mr_other
  DF$rv[DF$mr_true] <- "Mormon"

  # Catholics or Orthodox Christians ------------------------------
  co_r_num <- c(2, 10)
  co_o_num <- c(28, 123, 126, 143, 149, 183, 194)
  co_relig <- c_r[co_r_num]
  co_other <- c_o[co_o_num]
  DF$co_true <- (DF$relig %in% co_relig) | (DF$other %in% co_other)
  DF$rv[DF$co_true] <- "Catholic and Orthodox"

  # Jews ----------------------------------------------------------
  jw_r_num <- c(3)
  jw_relig <- c_r[jw_r_num]
  DF$jw_true <- DF$relig %in% jw_relig
  DF$rv[DF$jw_true] <- "Jew"

  # Other religions -----------------------------------------------
  DF$or_nonsp <- (DF$relig == c_r[5]) & !(DF$sp_pent |
                                          DF$sp_centchrist |
                                          DF$sp_fsg |
                                          DF$sp_jw |
                                          DF$sp_sda |
                                          DF$sp_ofund
                                          )
  or_r_num <- c(6:9, 12)
  or_o_num <- c(11, 74, 75, 80, 114, 136, 161, 163, 164, 192)
  or_relig <- c_r[or_r_num]
  or_other <- c_o[or_o_num]
  DF$or_true <- ((DF$relig %in% or_relig) |
                 (DF$other %in% or_other) |
                  DF$or_nonsp
                 )
  DF$rv[DF$or_true] <- "Other religions"

  # No religious identification -----------------------------------
  nr_r_num <- c(4)
  nr_relig <- c_r[nr_r_num]
  DF$nr_true <- DF$relig %in% nr_relig
  DF$rv[DF$nr_true] <- "No identification"

  # Missing values ------------------------------------------------
  # No Answer
  DF$na_relig <- DF$relig == c_r[99]
  DF$na_denom <- DF$denom == c_d[99]
  DF$na_rd <- DF$na_relig & DF$na_denom
  DF$rv[DF$na_rd] <- "No answer"

  # Don't know
  DF$dk_relig <- DF$relig == c_r[98]
  DF$rv[DF$dk_relig] <- "Don't know"


  ## FINISHING IT =================================================

  # Return the character vector with recoded religion
  return(DF$rv)
}





